File: layout.py
import itertools
import random

import numpy as np

class Layout:
    def __init__(self):
        self.nodes = []
        self.connections = []

        self.conditions = []
        self.spring_constant = 0.01
        self.repulsion_constant = 0.01
        self.pseudo_gravity_constant = 0.05

        self.num_control_points = 2
        # List of lists of nodes.
        # Each list corresponds to the virtual nodes within one connection.
        self.virtual_nodes = []

    def add_node(self):
        layout_node = LayoutNode()
        layout_node.x = random.uniform(0,1)
        layout_node.y = random.uniform(0,1)
        self.nodes.append(layout_node)

    def reset_nodes(self):
        for node in self.nodes:
            node.x = random.uniform(0,1)
            node.y = random.uniform(0,1)

        self.reset_edges()

    def add_connection(self, from_index, to_index):
        self.connections.append( (from_index,to_index) )
        control_points = self._gen_control_points(from_index, to_index)
        self.virtual_nodes.append(control_points)

    def reset_edges(self, num_control_points = None):
        if num_control_points is None:
            num_control_points = self.num_control_points

        self.virtual_nodes = [self._gen_control_points(from_index, to_index)
                              for (from_index, to_index) in self.connections]

    def _gen_control_points(self, from_index, to_index):
        initial = self.nodes[from_index].pos
        final = self.nodes[to_index].pos
        uniform = np.linspace(0, 1, self.num_control_points+2)[1:-1]
        mass = 0.1/self.num_control_points
        charge = 0.1/self.num_control_points
        return [LayoutNode(initial + num*(final-initial), mass=mass, charge=charge)
                for num in uniform]

    def _all_nodes(self, with_virtual=False):
        yield from self.nodes
        if with_virtual:
            yield from self._all_virtual_nodes()

    def _all_virtual_nodes(self):
        yield from itertools.chain(*self.virtual_nodes)

    def _all_node_pairs(self):
        yield from itertools.combinations(itertools.chain(
            self.nodes, self._all_virtual_nodes()), 2)

    def _connected_pairs(self):
        for (from_node, to_node) in self.connections:
            yield (self.nodes[from_node], self.nodes[to_node])

        for control_points in self.virtual_nodes:
            yield from zip(control_points[1:], control_points[:-1])

        for conn, control_points in zip(self.connections, self.virtual_nodes):
            if control_points:
                yield (self.nodes[conn[0]], control_points[0])
                yield (self.nodes[conn[1]], control_points[-1])

    def relax(self):
        # Electrostatic repulsion between pairs
        for (node_a, node_b) in self._all_node_pairs():
            disp = node_b.pos - node_a.pos
            dist2 = np.dot(disp, disp)
            if dist2 > 0:
                unit_vec = disp/np.sqrt(dist2)
                force = self.repulsion_constant * node_a.charge * node_b.charge * unit_vec / dist2
            else:
                force = np.array([1,0])

            node_a.pos -= force/node_a.mass
            node_b.pos += force/node_b.mass

        # Spring attraction between pairs
        for (node_a, node_b) in self._connected_pairs():
            disp = node_b.pos - node_a.pos
            force = -self.spring_constant * disp

            node_a.pos -= force/node_a.mass
            node_b.pos += force/node_b.mass

        # Pseudo-gravity, constant force toward zero
        for node in self._all_nodes(with_virtual=True):
            disp = node.pos
            node.pos -= self.pseudo_gravity_constant*len(self.nodes)/(1 + np.exp(-np.abs(disp)))

        # Apply conditions
        self._apply_conditions()

    def fix_x(self, node_index, rel_x):
        self.conditions.append(('fixed_x',node_index,rel_x))

    def fix_y(self, node_index, rel_y):
        self.conditions.append(('fixed_y',node_index,rel_y))

    def same_x(self, node_index_a, node_index_b):
        self.conditions.append(('same_x',node_index_a, node_index_b))

    def same_y(self, node_index_a, node_index_b):
        self.conditions.append(('same_y',node_index_a, node_index_b))

    def _apply_conditions(self):
        xmin = min(node.x for node in self._all_nodes())
        xmax = max(node.x for node in self._all_nodes())
        range_x = xmax - xmin

        ymin = min(node.y for node in self._all_nodes())
        ymax = max(node.y for node in self._all_nodes())
        range_y = ymax - ymin

        for condition in self.conditions:
            if condition[0] == 'fixed_x':
                self.nodes[condition[1]].x = xmin + range_x*condition[2]

            elif condition[0] == 'fixed_y':
                self.nodes[condition[1]].y = ymin + range_y*condition[2]

            elif condition[0] == 'same_x':
                new_x = sum(self.nodes[node_name].x for node_name in condition[1:])/len(condition[1:])
                for node_name in condition[1:]:
                    self.nodes[node_name].x = new_x

            elif condition[0] == 'same_y':
                new_y = sum(self.nodes[node_name].y for node_name in condition[1:])/len(condition[1:])
                for node_name in condition[1:]:
                    self.nodes[node_name].y = new_y

    def positions(self):
        """
        Returns the positions of all nodes and connections, ready to draw.
        There are three return values.
        1. node_pos, an array of positions of each node.
           nodes_pos[i][0] is the x position of the i-ith node.
           nodes_pos[i][1] is the y position of the i-ith node.

        2. connections, an array of the positions of the edges.
           connections[i][j][0] is the x position of the j-th point in the i-th connection
           connections[i][j][1] is the y position of the j-th point in the i-th connection
        """
        node_pos = np.array([node.pos for node in self.nodes])
        conn_origin = np.array([self.nodes[from_index].pos for (from_index, to_index) in self.connections])
        conn_dest = np.array([self.nodes[to_index].pos for (from_index, to_index) in self.connections])

        connections = []
        for i,_ in enumerate(self.connections):
            new_spline = self._control_points(i)
            connections.append(new_spline)
        connections = np.array(connections)

        return node_pos, connections

    def _control_points(self, i):
        """
        node_x_size, node_y_size are the size in real units of each node.

        Returns the spline representing the i-th edge.
        retval[j][0] is the x point of the j-th spline point.
        retval[j][1] is the y point of the j-th spline point.
        """
        from_index,to_index = self.connections[i]
        control_points = self.virtual_nodes[i]

        x = [p.x for p in control_points]
        x.insert(0, self.nodes[from_index].x)
        x.append(self.nodes[to_index].x)

        y = [p.y for p in control_points]
        y.insert(0, self.nodes[from_index].y)
        y.append(self.nodes[to_index].y)

        return np.array([x,y]).T



class LayoutNode:
    def __init__(self, pos=None, mass=1.0, charge=1.0):
        if pos is None:
            self.pos = np.array([0,0],dtype='float64')
        else:
            self.pos = np.array(pos, dtype='float64')

        self.mass = mass
        self.charge = charge

    @property
    def x(self):
        return self.pos[0]

    @x.setter
    def x(self, val):
        self.pos[0] = val

    @property
    def y(self):
        return self.pos[1]

    @y.setter
    def y(self, val):
        self.pos[1] = val

File: clayout/src/Layout.cc
#include "Layout.hh"

#include <algorithm>
#include <iostream>

Layout::Layout()
  : gen(std::random_device()()),
    spring_constant(0.01), repulsion_constant(0.01), pseudo_gravity_constant(0.01),
    num_control_points(2), num_spline_points(10), rel_node_size(0.05) { }

void Layout::add_node() {
  Node new_node;
  new_node.pos.X() = std::uniform_real_distribution<>(0,1)(gen);
  new_node.pos.Y() = std::uniform_real_distribution<>(0,1)(gen);
  new_node.mass = 1.0;
  new_node.charge = 1.0;
  nodes.push_back(new_node);
}

void Layout::add_connection(int from_index, int to_index) {
  Connection new_connection;
  new_connection.from_index = from_index;
  new_connection.to_index = to_index;
  gen_control_points(new_connection);
  connections.push_back(new_connection);
}

void Layout::gen_control_points(Connection& conn) {
  const GVector<2> initial = nodes[conn.from_index].pos;
  const GVector<2> final = nodes[conn.to_index].pos;
  conn.virtual_node_index = virtual_nodes.size();

  for(int i=0; i<num_control_points; i++) {
    Node control;
    control.pos = initial + (final-initial)*((i+1.0)/(num_control_points+1.0));
    control.mass = 0.1/num_control_points;
    control.charge = 0.1/num_control_points;
    virtual_nodes.push_back(control);
  }
}

void Layout::reset_node() {
  for(auto& node : nodes) {
    node.pos.X() = std::uniform_real_distribution<>(0,1)(gen);
    node.pos.Y() = std::uniform_real_distribution<>(0,1)(gen);
  }
  reset_edges();
}

void Layout::reset_edges() {
  virtual_nodes.clear();
  for(auto& conn : connections) {
    gen_control_points(conn);
  }
}

void Layout::relax() {
  for(unsigned int i=0; i<nodes.size(); i++) {
    for(unsigned int j=1; j<nodes.size(); j++) {
      electrostatic(nodes[i], nodes[j]);
    }
  }

  for(unsigned int i=0; i<virtual_nodes.size(); i++) {
    for(unsigned int j=1; j<virtual_nodes.size(); j++) {
      electrostatic(virtual_nodes[i], virtual_nodes[j]);
    }
  }

  for(auto& node : nodes) {
    for(auto& vnode : virtual_nodes) {
      electrostatic(node, vnode);
    }
  }


  for(auto& conn : connections) {
    spring(nodes[conn.from_index], nodes[conn.to_index]);
    if(num_control_points) {
      spring(nodes[conn.from_index], virtual_nodes[conn.virtual_node_index]);
      spring(nodes[conn.to_index], virtual_nodes[conn.virtual_node_index + num_control_points-1]);
    }
    for(int i=0; i+1<num_control_points; i++) {
      spring(virtual_nodes[conn.virtual_node_index],
             virtual_nodes[conn.virtual_node_index+1]);
    }
  }

  for(auto& node : nodes) {
    pseudo_gravity(node);
  }

  apply_constraints();
}

void Layout::electrostatic(Node& a, Node& b) {
  auto disp = b.pos - a.pos;
  auto dist2 = disp*disp;

  GVector<2> force;
  if(dist2 > 0) {
    force = repulsion_constant * a.charge * b.charge * disp.UnitVector() / dist2;
  } else {
    force = {1,0};
  }
  a.pos -= force/a.mass;
  b.pos += force/b.mass;
}

void Layout::spring(Node& a, Node& b) {
  auto disp = b.pos - a.pos;
  auto force = -spring_constant * disp;

  a.pos -= force/a.mass;
  b.pos += force/b.mass;
}

void Layout::pseudo_gravity(Node& a) {
  auto disp = a.pos;
  // auto force = pseudo_gravity_constant * nodes.size() *
  //   disp.UnitVector() / (1 + std::exp(-disp.Mag()));

  auto force = -pseudo_gravity_constant * nodes.size() * disp.UnitVector()
    / (1 + 1/disp.Mag());

  a.pos += force;
}

void Layout::apply_constraints() {
  // Fixed x positions
  auto x_elements = std::minmax_element(nodes.begin(), nodes.end(),
                                        [](const Node& a, const Node& b) { return a.pos.X() < b.pos.X(); } );
  double xmin = x_elements.first->pos.X();
  double xmax = x_elements.second->pos.X();
  double xrange = xmax - xmin;
  for(auto& con : fixed_x_pos) {
    nodes[con.first].pos.X() = xmin + xrange*con.second;
  }

  // Fixed y positions
  auto y_elements = std::minmax_element(nodes.begin(), nodes.end(),
                                        [](const Node& a, const Node& b) { return a.pos.Y() < b.pos.Y(); } );
  double ymin = y_elements.first->pos.Y();
  double ymax = y_elements.second->pos.Y();
  double yrange = ymax - ymin;
  for(auto& con : fixed_y_pos) {
    nodes[con.first].pos.Y() = ymin + yrange*con.second;
  }

  // Same x positions
  for(auto& con : same_x_pos) {
    double new_x = (nodes[con.first].pos.X() + nodes[con.second].pos.X())/2;
    nodes[con.first].pos.X() = new_x;
    nodes[con.second].pos.X() = new_x;
  }

  // Same y positions
  for(auto& con : same_y_pos) {
    double new_y = (nodes[con.first].pos.Y() + nodes[con.second].pos.Y())/2;
    nodes[con.first].pos.Y() = new_y;
    nodes[con.second].pos.Y() = new_y;
  }
}

DrawingPositions Layout::positions() const {
  DrawingPositions output;

  output.node_pos.reserve(nodes.size());
  for(auto& node : nodes) {
    output.node_pos.push_back(node.pos);
  }

  output.num_points_per_connection = num_control_points+2;
  output.connection_points.reserve((num_control_points+2) * connections.size());
  for(auto& conn : connections) {
    output.connection_points.push_back(nodes[conn.from_index].pos);
    for(int i=0; i<num_control_points; i++) {
      output.connection_points.push_back(virtual_nodes[conn.virtual_node_index+i].pos);
    }
    output.connection_points.push_back(nodes[conn.to_index].pos);
  }

  return output;
}
